java/com/google/re2j/Unicode.java

```diff
@@ -227,6 +229,99 @@ static int simpleFold(int r) {
     return toUpper(r);
   }
 
+  static int[] optimizedFoldOrbit(int r) {
+    if (r >= ORBIT_MIN_VALUE) {
+      return FOLD_MAP.get(r);
+    } else return null;
+  }
+
+  static int optimizedFoldNonOrbit(int r) {
+    int l = toLower(r);
+    if (l != r) {
+      return l;
+    }
+    return toUpper(r);
+  }
+  
+  private static final FoldMap FOLD_MAP = new FoldMap(UnicodeTables.CASE_ORBIT.length * 4);
+  private static final int ORBIT_MIN_VALUE = UnicodeTables.CASE_ORBIT[0][0];
+  
+  static {
+    for(int i = 0; i < UnicodeTables.CASE_ORBIT.length; i++) {
+      int r0 = UnicodeTables.CASE_ORBIT[i][0];
+      int[] folds = new int[0];
+      int r1 = r0;
+      while((r1 = simpleFold(r1)) != r0) {
+        folds = Arrays.copyOf(folds, folds.length + 1);
+        folds[folds.length - 1] = r1;
+      }
+      FOLD_MAP.put(r0, folds);
+    }
+  }
+
+  /*
+    associate a rune to it's unfolded case equivalent
+   */
+  private static class FoldMap {
+    private static final int MAX_DISTANCE = 4;
+    private final int[] keys;
+    private final int[][] values;
+    private final int mask;
+    private final int maxDistance;
+
+    FoldMap(int size) {
```

Comment:
FoldMap serves exactly one purpose.  It doesn't need a parameter, and it doesn't need to be separated from the logic that instantiates and populates it.